Add ERC: PermaLink Asset Bound Token (#999)
ethereum/ERCs: ERCS/erc-x.md

---
title: Non-Fungible PermaLink Asset Bound Token Standard
description: An interface for Non-Fungible PermaLink Asset Bound Tokens, also known as a NFABT.
author: Mihai Onila (@MihaiORO), Nick Zeman (@NickZCZ), Narcis Cotaie (@NarcisCRO)

> discussions-to: <[URL](https://ethereum-magicians.org/t/non-fungible-asset-bound-token/23175)>

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/non-fungible-asset-bound-token/23175
```

Author: Fixed

status: Draft
type: Standards Track

> category: <Core, Networking, Interface, or ERC> # Only required for Standards Track. Otherwise, remove this field.

Reviewer: ```suggestion
category: ERC
```

Author: Fixed

> created: <date created on, in ISO 8601 (yyyy-mm-dd) format>

Reviewer: Please include the date this EIP proposal was first created.

Author: Fixed

requires: <EIP number(s)> # Only required when you reference an EIP in the `Specification` section. Otherwise, remove this field.
---

> <!--
>   READ EIP-1 (https://eips.ethereum.org/EIPS/eip-1) BEFORE USING THIS TEMPLATE!
> 
>   This is the suggested template for new EIPs. After you have filled in the requisite fields, please delete these comments.
> 
>   Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`.
> 
>   The title should be 44 characters or less. It should not repeat the EIP number in title, irrespective of the category.
> 
>   TODO: Remove this comment before submitting
> -->

Reviewer: And all of this should go away.
```suggestion

```

Author: Fixed

## Abstract

A standard interface for Non-Fungible PermaLink Asset Bound Tokens (PermaLink-ABT/s), a subset of the more general Asset Bound Tokens (ABT/s).

The following standard allows for the implementation of a standard API for tokens within smart contracts and provides mirrored information of a specific smart contract through the ‘assetBoundContract’ function. Mirrored information consists of ‘ownerOf’, ‘tokenID’, ‘totalSupply’, as well as ‘balanceOf’. This in conjunction with blocking the ability to use basic transfer and approve functionality makes 2 tokens from different smart contracts interlocked. As the asset cannot be transferred in the traditional sense, being bound to a specific asset within a specific contract and maintaining corresponding information and movements creates an ABT. 

As the asset is bound to another asset having specified mirrored information, a ‘reveal’ function replaces the mint function commonly seen. The total supply of all the tokens from the ABT smart contract already exists.

> We considered ABTs being used by every individual who wishes to link an additional non-fungible asset to an already existing non-fungible asset. ABTs allow an infinite amount of tokens to be bound together, allowing them to work symbiotically rather than splintering and separating. Remaining interlocked and asset bound removes the idea that 2 smart contracts released by the same or different creators at the same or different times are competing with one another.

Reviewer: Your abstract contains both technical specifications and rationale, which should be separated into appropriate sections. The technical details should go in the Specification section, and the reasoning behind design choices should go in the Rationale section.

Author: Hopefully fixed :)

> <!--
>   The Abstract is a multi-sentence (short paragraph) technical summary. This should be a very terse and human-readable version of the specification section. Someone should be able to read only the abstract to get the gist of what this specification does.
> 
>   TODO: Remove this comment before submitting
> -->

Reviewer: ```suggestion

```

ABTs streamline inventory management by allowing dynamic asset binding. Instead of creating redundant smart contracts, manufacturers can link tokens representing various stages of the supply chain, maintaining historical data while ensuring seamless tracking and updates.

4. **Addressing NFT Fragmentation:** NFTs have traditionally been associated with digital collectibles and art. However, many projects deploy secondary smart contracts to evolve their collections, inadvertently causing liquidity fragmentation. Owners often sell assets from the original collection to acquire newer ones, leading to value dilution and lower overall market confidence.
ABTs solve this by allowing secondary collections to complement rather than compete with the original. Bound assets enhance the primary asset’s value without necessitating a separate, competing ecosystem. This structure retains liquidity within the original collection and sustains its market metrics, benefiting both creators and collectors.

> 5. **New Opprutunites for Creators:** ABTs empower both asset owners and creators by enabling an open secondary market for existing tokens. Artists, for instance, can create and bind assets to existing NFTs without requiring permission from the original contract owner. This facilitates new revenue streams, such as artists being paid on consignment for augmenting or adding onto existing assets. Owners benefit as well, as additional assets increase the inherent value of their holdings, especially in instances of collaboration between established projects or reputable artists.

Reviewer: ```suggestion
5. **New Opportunities for Creators:** ABTs empower both asset owners and creators by enabling an open secondary market for existing tokens. Artists, for instance, can create and bind assets to existing NFTs without requiring permission from the original contract owner. This facilitates new revenue streams, such as artists being paid on consignment for augmenting or adding onto existing assets. Owners benefit as well, as additional assets increase the inherent value of their holdings, especially in instances of collaboration between established projects or reputable artists.
```

Author: Fixed

<!--
  This section is optional.

  The motivation section should include a description of any nontrivial problems the EIP solves. It should not describe how the EIP solves those problems, unless it is not immediately obvious. It should not describe why the EIP should be made into a standard, unless it is not immediately obvious.

  With a few exceptions, external links are not allowed. If you feel that a particular resource would demonstrate a compelling case for your EIP, then save it as a printer-friendly PDF, put it in the assets folder, and link to that copy.

  TODO: Remove this comment before submitting

> -->

Reviewer: Remove the template comment.

---
title: PermaLink Asset Bound Token Standard
description: An interface for PermaLink Asset Bound Tokens, also known as a PermaLink-ABTs.
author: Mihai Onila (@MihaiORO), Nick Zeman (@NickZCZ), Narcis Cotaie (@NarcisCRO)
discussions-to: https://ethereum-magicians.org/t/non-fungible-asset-bound-token/23175
status: Draft
type: Standards Track
category: ERC
created: 2025-04-01

> requires: 721, 721Enumerable

Reviewer 2: ```suggestion
requires: 721
```

Please find the EIP number for `EIP721Enumerable` and use that here.

---

### Abstract

> This standard introduces a subclass of tokens known as **PermaLink Asset Bound Tokens (PermaLink-ABTs)** a specific implementation of the broader **Asset Bound Token (ABT)** concept. ABTs establish a novel ownership paradigm where **an asset can own another asset**, enabling composable, nested, and portfolio-like token structures that evolve together over time.

Reviewer 2: ```suggestion
This standard introduces a subclass of tokens known as **PermaLink Asset Bound Tokens (PermaLink-ABTs)**. They are a specific implementation of the broader **Asset Bound Token (ABT)** concept. ABTs establish a novel ownership paradigm where **an asset can own another asset**, enabling composable, nested, and portfolio-like token structures that evolve together over time.
```

> PermaLink-ABTs implement a permanent binding mechanism where a token in one smart contract is irreversibly linked to a token in another contract. These links mirror key state data such as `ownerOf`, `tokenId`, `totalSupply`, and `balanceOf` using the `assetBoundContract` interface. Traditional token transfer and approval functions are disabled to enforce immutability and structural cohesion between bound assets.

Reviewer 2: ```suggestion
PermaLink-ABTs implement a permanent binding mechanism where a token in one smart contract is irreversibly linked to a token in another contract. These links mirror key state data such as `ownerOf`, `tokenId`, `totalSupply`, and `balanceOf` using the `assetBoundContract` interface. Traditional token transfer and approval functions are omitted to enforce immutability and structural cohesion between bound assets.
```

Saying "omitted" is a bit more clear, since you don't actually include `transfer` (and friends) in your interface above.

By encouraging asset composability over competition, PermaLink-ABTs introduce a dynamic, future-proof model for on-chain asset evolution.

## Motivation

Traditional ownership models on Ethereum are inherently limited—only externally owned accounts (EOAs) or smart contracts can own blockchain assets. This creates rigidity, especially as the blockchain ecosystem expands to include digital identities, tokenized real-world assets (RWAs), and NFTs. Current smart contracts are static in nature, meaning once deployed, they cannot adapt to evolving systems, unforeseen use cases, or new integration layers. There is a need for a more flexible and modular ownership model that allows for dynamic interactions between assets.

This standard proposes **Asset Bound Tokens (ABTs)** as a solution to this challenge. ABTs allow one token to be permanently bound to another across contracts, creating a dynamic, flexible, and composable ownership model. By enabling tokens to move and evolve together, ABTs pave the way for new possibilities in the on-chain economy. The following use cases illustrate why ABTs are needed:

1. **On-Chain Identity Systems**  

>  Governments and institutions worldwide are piloting or implementing blockchain-based identity systems—digital passports, national IDs, and verifiable credentials such as with the EBSI initiative. These systems often require credentials to be linked across multiple registries (e.g., healthcare, banking, voting). ABTs enable the binding of identity-linked tokens into a cohesive unit, so they move together instead of requiring manual coordination and transfers. This ensures that identity-linked assets remain interconnected and dynamic, making it easier to manage and update linked data as users interact with various systems.

Reviewer 2: > EBSI initiative

Could you expand this abbreviation?

5. **New Opportunities for Creators**  
   ABTs empower creators to build on top of existing assets permissionlessly—without needing ownership of or permission from the original smart contract. This enables a new wave of creative expression, where artists can augment and enhance NFTs (e.g., adding new visuals, audio, or interactive layers) and collaborate on existing collections. Such contributions can generate new revenue streams through shared royalties, consignment, or collaborative upgrades. Owners benefit as well, since bound enhancements can increase the inherent value of their holdings, particularly in projects involving established creators or cross-collection collaborations.

In essence, ABTs introduce a framework where tokens are linked rather than owned. This allows for dynamic and evolving asset systems, where assets move together in harmony. If a binding token moves, all associated ABTs move with it, ensuring seamless updates and reducing the need for manual transfers. This innovation transforms traditional smart contracts from static repositories into living, evolving systems capable of adapting to changing use cases, technologies, and business models.

> The **PermaLink-ABTs** implementation takes the ABT model further by enforcing a permanent binding between one token and another—whether it’s another ABT, NFT or NFKBT. This permanent binding ensures that tokens can be transferred as a single unit, reducing complexity and gas fees. Instead of relying on traditional minting, PermaLink-ABTs use a `reveal` mechanism, activating tokens from a predefined supply. This reduces gas costs and encourages efficient linking of assets, enabling greater composability across multiple sectors.

Reviewer 2: ```suggestion
The **PermaLink-ABTs** implementation takes the ABT model further by enforcing a permanent binding between one token and another—whether it's another ABT, NFT or NFKBT. This permanent binding ensures that tokens can be transferred as a single unit, reducing complexity and gas fees. Instead of relying on traditional minting, PermaLink-ABTs use a `reveal` mechanism, activating tokens from a predefined supply. This reduces gas costs and encourages efficient linking of assets, enabling greater composability across multiple sectors.
```

PermaLink-ABTs consolidate asset value by allowing multiple subordinate tokens to be linked to a single binding token, providing enhanced composability and reducing fragmentation. By requiring only the binding token to be transferred, all associated assets move in sync, making it easier to manage portfolios and move groups of assets together. This approach fosters collaboration, value accrual, and compatibility across ecosystems, whether for digital identities, RWAs, NFTs, or other on-chain assets.

## Specification

### `IABT` (Token Interface)

**NOTES**:

- The following specifications use syntax from Solidity `0.8.27` (or above)
- Callers MUST handle `false` from `returns (bool success)`. Callers MUST NOT assume that `false` is never returned!

```solidity
interface IABT {
    function ownerOf(uint256 tokenId) external view returns (address);
    function tokenExists(uint256 tokenId) external view returns (bool);
    function totalSupply() external view returns (uint256);
    function balanceOf(address owner) external view returns (uint256);
}
```

> ### `ABT` (Token Contract)

Reviewer 2: Implementations of the standard should be in the Reference Implementation section, not in the specification itself.

        uint256
    ) public pure override(ERC721, IERC721) {
        revert("ABT: Transfers not allowed");
    }

    function safeTransferFrom(
        address,
        address,
        uint256,
        bytes memory
    ) public pure override(ERC721, IERC721) {
        revert("ABT: Transfers not allowed");
    }
}
```

### **Interface functions**

The functions detailed below MUST be implemented.

#### `ownerOf` function

Returns the owner of the NFT specified by the `tokenId`. Will read from the `assetBoundContract` the owner and return it.

```solidity
function ownerOf(uint256 tokenId) external view returns (address);
```

#### `tokenExists` function

Returns true if the owner of the token is not the zero address (`0x0`)

```solidity
function tokenExists(uint256 tokenId) external view returns (bool);
```

#### `totalSupply` function

Gets the total amount of tokens stored by the assetBoundContract

```solidity
function totalSupply() external view returns (uint256);
```

#### `balanceOf` function

Returns the number of NFTs in the assetBoundContract that an owner has.

```solidity
function balanceOf(address owner) external view returns (uint256);
```

## Rationale

> The design of PermaLink-ABTs centers around the goal of enabling permanent token binding while optimizing for gas efficiency, composability, and secure ownership structures. We chose to implement a `reveal` function in place of a traditional `mint` function to reduce gas costs and simplify on-chain state changes. Unlike minting, which creates tokens at runtime and incurs higher gas fees, the `reveal` function maps pre-allocated tokens stored in an array or mapping. This allows tokens to be activated on demand without the overhead of dynamic token creation. As a result, token issuers can prepare and store an entire supply in advance, with users later revealing and binding tokens when needed. This approach aligns with the use case of portfolio binding and asset hierarchies, where large numbers of tokens may need to be activated and bound efficiently. 

Reviewer 2: Your specification doesn't define a `reveal` function. That said, we generally recommend _not_ specifying `mint`/`burn`-type functions because they're usually admin-only. The creator of a contract will know how to call the admin functions, so they don't need to be part of the standard (unless you expect it to be common for _anyone_ to be minting/burning/revealing).

> We adopted the `assetBoundContract` interface to mirror essential metadata such as `ownerOf`, `tokenId`, `totalSupply`, and `balanceOf` from the binding token’s contract. This ensures that PermaLink-ABTs remain synchronized with the asset they are bound to, without duplicating logic or requiring manual updates. The mirroring also ensures traceability and visibility across contracts, allowing observers and off-chain systems to reliably interpret the token relationship. To preserve the permanent nature of the bond, standard `transfer` and `approve` methods are disabled. This immutability guarantees that PermaLink-ABTs cannot be separated from their bound asset once revealed. If the primary token moves, all attached PermaLink-ABTs move with it. This behavior supports composability, value aggregation, and consistent ownership logic.

Reviewer 2: ```suggestion
We adopted the `assetBoundContract` interface to mirror essential metadata such as `ownerOf`, `tokenId`, `totalSupply`, and `balanceOf` from the binding token's contract. This ensures that PermaLink-ABTs remain synchronized with the asset they are bound to, without duplicating logic or requiring manual updates. The mirroring also ensures traceability and visibility across contracts, allowing observers and off-chain systems to reliably interpret the token relationship. To preserve the permanent nature of the bond, standard `transfer` and `approve` methods are disabled. This immutability guarantees that PermaLink-ABTs cannot be separated from their bound asset once revealed. If the primary token moves, all attached PermaLink-ABTs move with it. This behavior supports composability, value aggregation, and consistent ownership logic.
```

PermaLink-ABTs enforce strict one-way binding with immutable relationships, making them especially suitable for use cases like identity systems, real-world asset (RWA) structures, and portfolio-locked NFTs. They act as permanently attached extensions to existing tokens, reducing complexity and avoiding redundant contract logic. This approach also provides a cleaner and more secure way to augment existing assets while maintaining compatibility across various blockchain use cases. This standard is intentionally minimal to ensure wide compatibility and flexibility. Developers can extend the base logic for specialized use cases, such as embedding royalty splits, upgrade paths, or linking to dynamic data feeds, without altering the underlying PermaLink mechanism.

## Backwards Compatibility

> <!--
> 
>   This section is optional.
> 
>   All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.
> 
>   The current placeholder is acceptable for a draft.
> 
>   TODO: Remove this comment before submitting
> -->
> 

Reviewer 2: ```suggestion
```

No backward compatibility issues found.

> ## Test Cases
> 
> <!--
>   This section is optional for non-Core EIPs.
> 
>   The Test Cases section should include expected input/output pairs, but may include a succinct set of executable tests. It should not include project build files. No new requirements may be introduced here (meaning an implementation following only the Specification section should pass all tests here.)
>   If the test suite is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`. External links will not be allowed
> 
>   TODO: Remove this comment before submitting
> -->
> 

Reviewer 2: ```suggestion
```

## Reference Implementation

<!--
  This section is optional.

  The Reference Implementation section should include a minimal implementation that assists in understanding or implementing this specification. It should not include project build files. The reference implementation is not a replacement for the Specification section, and the proposal should still be understandable without it.
  If the reference implementation is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`. External links will not be allowed.

  TODO: Remove this comment before submitting

> -->

Reviewer 2: Replace this with your implementation above.

## Security Considerations

PermaLink-ABTs are linked to another non-fungible token. If an individual loses access to this token—what we call the **binding token**—they also lose access to all PermaLink-ABTs that have been bound to it. This introduces a critical security consideration: the entire value of bound assets depends on the integrity and availability of the binding token.

> To mitigate this risk, we strongly recommend the use of standards like [ERC-6809](https://eips.ethereum.org/EIPS/eip-6809), a **Non-Fungible Key Bound Token**, which introduces on-chain two-factor authentication (2FA). ERC-6809 allows a user to bind sensitive tokens (like PermaLink-ABTs) to a secured identity layer, complete with recovery mechanisms. In the event that a user loses access to their original wallet or interacts with a malicious contract, ERC-6809 provides a safeFallback function to re-establish control.

Reviewer 2: ```suggestion
To mitigate this risk, we strongly recommend the use of standards like [ERC-6809](./eip-6809.md), a **Non-Fungible Key Bound Token**, which introduces on-chain two-factor authentication (2FA). ERC-6809 allows a user to bind sensitive tokens (like PermaLink-ABTs) to a secured identity layer, complete with recovery mechanisms. In the event that a user loses access to their original wallet or interacts with a malicious contract, ERC-6809 provides a `safeFallback` function to re-establish control.
```

In essence, all of the security guarantees of ERC-6809 extend to any PermaLink-ABTs bound to it. This layered security model not only protects against loss but also ensures recoverability and long-term viability for high-value bound assets. It is strongly encouraged that developers implementing PermaLink-ABTs integrate this or similar standards to provide a robust security foundation for users.

> Needs discussion.

Reviewer 2: ```suggestion
```

---
eip: 7929
title: PermaLink Asset Bound Token Standard

> description: An interface for PermaLink Asset Bound Tokens, also known as a PermaLink-ABTs.

Reviewer 2: Your description restates a bit too much of the title. You want to use this field to expand on the ideas introduced in your title instead of just repeating the same information.

```suggestion
description: Asset-bound Tokens (ABT) permanently bound to another token, known as PermaLink-ABTs.
```

How's that?

The **PermaLink-ABTs** implementation takes the ABT model further by enforcing a permanent binding between one token and another—whether it's another ABT, NFT or NFKBT. This permanent binding ensures that tokens can be transferred as a single unit, reducing complexity and gas fees. Instead of relying on traditional minting, PermaLink-ABTs use a `reveal` mechanism, activating tokens from a predefined supply. This reduces gas costs and encourages efficient linking of assets, enabling greater composability across multiple sectors.

PermaLink-ABTs consolidate asset value by allowing multiple subordinate tokens to be linked to a single binding token, providing enhanced composability and reducing fragmentation. By requiring only the binding token to be transferred, all associated assets move in sync, making it easier to manage portfolios and move groups of assets together. This approach fosters collaboration, value accrual, and compatibility across ecosystems, whether for digital identities, RWAs, NFTs, or other on-chain assets.

## Specification

### `IABT` (Token Interface)

**NOTES**:

- The following specifications use syntax from Solidity `0.8.27` (or above)

> - Callers MUST handle `false` from `returns (bool success)`. Callers MUST NOT assume that `false` is never returned!

Reviewer 2: Maybe need to move this line.

```solidity
interface IABT {
    event AssetBoundContractSet(address assetBoundContract);
    event TokenRevealed(uint256 tokenId);

    function ownerOf(uint256 tokenId) external view returns (address);
    function tokenExists(uint256 tokenId) external view returns (bool);
    function totalSupply() external view returns (uint256);
    function balanceOf(address owner) external view returns (uint256);

    function reveal(uint256[] calldata tokenIds) external payable;
}
```

> ### **Interface events**

Reviewer 2: ```suggestion
### Events
```

#### `AssetBoundContractSet` event

Emitted when the contract is deployed and bound to `assetBoundContract`

```solidity
event AssetBoundContractSet(address assetBoundContract);
```

#### `TokenRevealed` event

Emitted when the `tokenId` is revealed

```solidity
event TokenRevealed(uint256 tokenId);
```

### **Interface functions**

The functions detailed below MUST be implemented.

> #### `ownerOf` function

Reviewer 2: ```suggestion
#### `ownerOf`
```

Returns the owner of the NFT specified by the `tokenId`. Will read from the `assetBoundContract` the owner and return it.

```solidity
function ownerOf(uint256 tokenId) external view returns (address);
```

> #### `tokenExists` function

Reviewer 2: ```suggestion
#### `tokenExists`
```

> Returns true if the owner of the token is not the zero address (`0x0`)

Reviewer 2: This is a very strict requirement. So the zero address is never allowed to own tokens? Would it make more sense to define this more vaguely in case something comes along in the future where the zero address is useful for some reason?

```solidity
function tokenExists(uint256 tokenId) external view returns (bool);
```

#### `totalSupply` function

Gets the total amount of tokens stored by the assetBoundContract

```solidity
function totalSupply() external view returns (uint256);
```

#### `balanceOf` function

Returns the number of NFTs in the assetBoundContract that an owner has.

```solidity
function balanceOf(address owner) external view returns (uint256);
```

### `reveal` function

An optional `reveal` function MAY be implemented to allow pre-allocated tokens to be activated on demand.  
This method reduces gas consumption compared to traditional minting and simplifies token activation mechanics.

```solidity
function reveal(uint256[] calldata tokenIds) external payable;

> ```

Reviewer 2: It might make sense to define this in a separate interface. This approach has the benefit of being more compatible with ERC-165's `supportsInterface` introspection.
